qubes/storage/lvm.py

```diff
@@ -0,0 +1,344 @@
+        result = p.communicate()[0]
+        volumes = []
+        for line in result.splitlines():
+            if not line.strip():
+                continue
+            vid, atr = line.strip().split(' ')
+            config = {
+                'pool': self.name,
+                'vid': vid,
+                'name': vid,
+                'volume_group': self.volume_group,
+                'rw': atr[1] == 'w',
+            }
+            volumes += [ThinVolume(**config)]
+        return volumes
+
+    def _reset_volume(self, volume):
+        ''' Resets a volatile volume '''
+        assert volume.volume_type == 'volatile', \
+            'Expected a volatile volume, but got {!r}'.format(volume)
+        self.log.debug('Resetting volatile ' + volume.vid)
+        cmd = ['remove', volume.vid]
+        qubes_lvm(cmd, self.log)
+        cmd = ['create', self._pool_id, volume.vid.split('/')[1],
+               str(volume.size)]
+        qubes_lvm(cmd, self.log)
+
+class ThinVolume(qubes.storage.Volume):
+    ''' Default LVM thin volume implementation
+    '''  # pylint: disable=too-few-public-methods
+
+    def __init__(self, volume_group, **kwargs):
+        self.volume_group = volume_group
+        super(ThinVolume, self).__init__(**kwargs)
+
+        if self.snap_on_start and self.source is None:
+            msg = "snap_on_start specified on {!r} but no volume source set"
+            msg = msg.format(self.name)
+            raise qubes.storage.StoragePoolException(msg)
+        elif not self.snap_on_start and self.source is not None:
+            msg = "source specified on {!r} but no snap_on_start set"
+            msg = msg.format(self.name)
+            raise qubes.storage.StoragePoolException(msg)
+
+        self.path = '/dev/' + self.vid
+        if not self._is_volatile:
+            self._vid_snap = self.vid + '-snap'
+
+    @property
+    def revisions(self):
+        path = self.path + '-back'
+        if os.path.exists(path):
+            seconds = os.path.getctime(path)
+            iso_date = qubes.storage.isodate(seconds).split('.', 1)[0]
+            return {iso_date: path}
+        return {}
+
+    @property
+    def _is_origin(self):
+        return not self.snap_on_start and self.save_on_stop
+
+    @property
+    def _is_origin_snapshot(self):
+        return self.snap_on_start and self.save_on_stop
+
+    @property
+    def _is_snapshot(self):
+        return self.snap_on_start and not self.save_on_stop
+
+    @property
+    def _is_volatile(self):
+        return not self.snap_on_start and not self.save_on_stop
+
+def pool_exists(pool_id):
+    ''' Return true if pool exists '''
+    cmd = ['pool', pool_id]
+    return qubes_lvm(cmd)
+
+
+def qubes_lvm(cmd, log=logging.getLogger('qube.storage.lvm')):
+    ''' Call :program:`qubes-lvm` to execute an LVM operation '''
+    # TODO Refactor this ones the udev groups gets fixed and we don't need root
+    # for operations on lvm devices
+    cmd = ['sudo', 'qubes-lvm'] + cmd
```

Comment:
> The individual qubes-lvm subcommands should call sudo lvm ... if needed, we shouldn't elevate parts of core to root priviledges

The reason for separating the lvm part in `qubes-lvm` was to not rely on `sudo lv*` commands, but to reuse the lvm2 library. In such case your executable needs to run as root. This executable is `qubes-lvm`. Because at the end everything worked out, I have not migrated yet all the 'lv *' calls in `qubes_lvm`.

In future I need to figure out the udev rules to make the lvm devices accessible from the `qubes` group.